File: .agents/skills/crypto-daily/scripts/notion_publish.py
#!/usr/bin/env python3
"""
notion_publish.py — Publish a Markdown report to a Notion page.

Usage:
  python3 notion_publish.py \
    --token ntn_xxx \
    --parent <parent_page_id> \
    --title "📊 Crypto Daily — 2026-06-23" \
    --report research/crypto-portfolio-2026-06-23.md

Prints the Notion page URL on success, exits non-zero on failure.
"""

import argparse
import json
import re
import sys

NOTION_VERSION = "2022-06-28"
API_BASE = "https://api.notion.com/v1"

try:
    import requests
    def _notion_request(method: str, endpoint: str, token: str, body: dict) -> dict:
        url = f"{API_BASE}{endpoint}"
        fn = requests.post if method == "POST" else requests.patch
        r = fn(url, json=body, headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
            "Notion-Version": NOTION_VERSION,
        })
        if not r.ok:
            print(f"❌ Notion API error {r.status_code}: {r.text}", file=sys.stderr)
            sys.exit(1)
        return r.json()
    def api_post(endpoint, token, body): return _notion_request("POST",  endpoint, token, body)
    def api_patch(endpoint, token, body): return _notion_request("PATCH", endpoint, token, body)
except ImportError:
    import urllib.request, urllib.error
    def _notion_request(method, endpoint, token, body):
        url = f"{API_BASE}{endpoint}"
        data = json.dumps(body, ensure_ascii=False).encode("utf-8")
        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("Authorization", f"Bearer {token}")
        req.add_header("Content-Type", "application/json; charset=utf-8")
        req.add_header("Notion-Version", NOTION_VERSION)
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            print(f"❌ Notion API error {e.code}: {e.read().decode()}", file=sys.stderr)
            sys.exit(1)
    def api_post(endpoint, token, body): return _notion_request("POST",  endpoint, token, body)
    def api_patch(endpoint, token, body): return _notion_request("PATCH", endpoint, token, body)


def md_to_notion_blocks(md: str) -> list:
    """
    Convert Markdown to Notion block objects.
    Handles: headings (# ## ###), bullets (- *), code fences, bold, paragraphs.
    Long text is split into 2000-char chunks (Notion rich_text limit).
    """
    blocks = []
    lines = md.split("\n")
    in_code = False
    code_lang = ""
    code_lines: list[str] = []

    def rich_text(text: str) -> list[dict]:
        """Split text into ≤2000-char rich_text chunks."""
        chunks = []
        while text:
            chunk, text = text[:2000], text[2000:]
            chunks.append({"type": "text", "text": {"content": chunk}})
        return chunks or [{"type": "text", "text": {"content": ""}}]

    def flush_code():
        nonlocal in_code, code_lang, code_lines
        content = "\n".join(code_lines)
        # Notion code blocks have a 2000-char limit per block — split if needed
        while content:
            chunk, content = content[:1990], content[1990:]
            blocks.append({
                "object": "block",
                "type": "code",
                "code": {
                    "rich_text": [{"type": "text", "text": {"content": chunk}}],
                    "language": code_lang or "plain text",
                },
            })
        in_code = False
        code_lang = ""
        code_lines = []

    for line in lines:
        # Code fence toggle
        if line.startswith("```"):
            if not in_code:
                in_code = True
                code_lang = line[3:].strip() or "plain text"
            else:
                flush_code()
            continue

        if in_code:
            code_lines.append(line)
            continue

        # Headings
        if line.startswith("### "):
            blocks.append({
                "object": "block", "type": "heading_3",
                "heading_3": {"rich_text": rich_text(line[4:].strip())},
            })
        elif line.startswith("## "):
            blocks.append({
                "object": "block", "type": "heading_2",
                "heading_2": {"rich_text": rich_text(line[3:].strip())},
            })
        elif line.startswith("# "):
            blocks.append({
                "object": "block", "type": "heading_1",
                "heading_1": {"rich_text": rich_text(line[2:].strip())},
            })
        # Bullets
        elif re.match(r"^[-*]\s+", line):
            text = re.sub(r"^[-*]\s+", "", line)
            blocks.append({
                "object": "block", "type": "bulleted_list_item",
                "bulleted_list_item": {"rich_text": rich_text(text)},
            })
        # Table rows → paragraph (Notion table blocks are complex; keep as text)
        elif line.startswith("|"):
            blocks.append({
                "object": "block", "type": "paragraph",
                "paragraph": {"rich_text": rich_text(line)},
            })
        # Horizontal rule → divider
        elif re.match(r"^---+$", line.strip()):
            blocks.append({"object": "block", "type": "divider", "divider": {}})
        # Blank line → skip (don't add empty paragraphs)
        elif line.strip() == "":
            continue
        # Regular paragraph
        else:
            blocks.append({
                "object": "block", "type": "paragraph",
                "paragraph": {"rich_text": rich_text(line)},
            })

    if in_code:
        flush_code()

    return blocks


def main():
    parser = argparse.ArgumentParser(description="Publish Markdown report to Notion")
    parser.add_argument("--token", required=True, help="Notion API token")
    parser.add_argument("--parent", required=True, help="Parent page ID (32-char hex)")
    parser.add_argument("--title", required=True, help="Page title")
    parser.add_argument("--report", required=True, help="Path to Markdown report file")
    args = parser.parse_args()

    # Read report
    try:
        with open(args.report) as f:
            md = f.read()
    except FileNotFoundError:
        print(f"❌ Report file not found: {args.report}", file=sys.stderr)
        sys.exit(1)

    # Notion API: max 100 blocks per request — batch
    all_blocks = md_to_notion_blocks(md)

    # Normalise parent ID (strip hyphens if user pasted URL form)
    parent_id = args.parent.replace("-", "")
    if len(parent_id) == 32:
        parent_id = f"{parent_id[:8]}-{parent_id[8:12]}-{parent_id[12:16]}-{parent_id[16:20]}-{parent_id[20:]}"

    # Create the page (first batch up to 100 blocks)
    first_batch = all_blocks[:100]
    payload = {
        "parent": {"type": "page_id", "page_id": parent_id},
        "properties": {
            "title": {
                "title": [{"type": "text", "text": {"content": args.title}}]
            }
        },
        "children": [],  # create empty; append blocks via PATCH
    }

    result = api_post("/pages", args.token, payload)
    page_id = result["id"]
    page_url = result.get("url", f"https://app.notion.com/p/{page_id.replace('-', '')}")

    # Append all blocks in batches of 50 via PATCH
    for i in range(0, len(all_blocks), 50):
        batch = all_blocks[i:i+50]
        api_patch(f"/blocks/{page_id}/children", args.token, {"children": batch})

    print(f"✅ Notion page created: {page_url}")
    return page_url


if __name__ == "__main__":
    main()

File: .agents/scripts/publish/notion_publish.ts
#!/usr/bin/env bun
/**
 * Publish a markdown file as a Notion child page.
 *
 * Usage:
 *   bun .agents/scripts/publish/notion_publish.ts \
 *     --file research/foo.md --parent <32-hex-page-id> --title "My Title"
 *
 * Token resolution order: $NOTION_TOKEN, then the `notion` MCP entry in
 * ~/.config/opencode/opencode.json. Never printed.
 *
 * Markdown tables are emitted as `code` blocks: Notion's table block requires a
 * fixed column count declared up front and nested table_row children, which
 * breaks on the ragged tables these reports contain. A monospace code block
 * preserves alignment and is lossless, which matters more than native styling
 * for an audit artifact.
 */

const NOTION_VERSION = "2022-06-28";
const MAX_BLOCKS_PER_CALL = 100;
const MAX_RICH_TEXT = 1900; // Notion hard limit is 2000; leave headroom.

function arg(name: string): string | undefined {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 ? process.argv[i + 1] : undefined;
}

async function resolveToken(): Promise<string> {
  if (process.env.NOTION_TOKEN) return process.env.NOTION_TOKEN;
  const cfgPath = `${process.env.HOME}/.config/opencode/opencode.json`;
  const raw = await Bun.file(cfgPath).text();
  const m = raw.match(/"NOTION_TOKEN"\s*:\s*"([^"]+)"/);
  if (!m) throw new Error("NOTION_TOKEN not found in env or opencode.json");
  return m[1];
}

/** Split any string into <=MAX_RICH_TEXT chunks so Notion never 400s. */
function richText(s: string) {
  const out: any[] = [];
  for (let i = 0; i < Math.max(s.length, 1); i += MAX_RICH_TEXT) {
    out.push({ type: "text", text: { content: s.slice(i, i + MAX_RICH_TEXT) } });
  }
  return out;
}

function para(s: string) {
  return { object: "block", type: "paragraph", paragraph: { rich_text: richText(s) } };
}

function heading(level: number, s: string) {
  const t = `heading_${Math.min(level, 3)}` as const;
  return { object: "block", type: t, [t]: { rich_text: richText(s) } } as any;
}

function codeBlock(s: string, lang = "plain text") {
  return {
    object: "block",
    type: "code",
    code: { rich_text: richText(s), language: lang },
  };
}

/** Strip markdown emphasis so Notion plain text stays readable. */
function clean(s: string): string {
  return s
    .replace(/\*\*(.+?)\*\*/g, "$1")
    .replace(/`(.+?)`/g, "$1")
    .replace(/\[([^\]]+)\]\(([^)]+)\)/g, "$1 ($2)")
    .trim();
}

function mdToBlocks(md: string): any[] {
  const lines = md.split("\n");
  const blocks: any[] = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];
    const t = line.trim();

    if (!t) { i++; continue; }

    // fenced code
    if (t.startsWith("```")) {
      const buf: string[] = [];
      i++;
      while (i < lines.length && !lines[i].trim().startsWith("```")) buf.push(lines[i++]);
      i++;
      blocks.push(codeBlock(buf.join("\n") || " "));
      continue;
    }

    // table -> code block (see header comment for why)
    if (t.startsWith("|")) {
      const buf: string[] = [];
      while (i < lines.length && lines[i].trim().startsWith("|")) {
        buf.push(clean(lines[i].trim()));
        i++;
      }
      blocks.push(codeBlock(buf.join("\n")));
      continue;
    }

    if (/^---+$/.test(t)) {
      blocks.push({ object: "block", type: "divider", divider: {} });
      i++;
      continue;
    }

    const h = t.match(/^(#{1,6})\s+(.*)$/);
    if (h) {
      blocks.push(heading(h[1].length, clean(h[2])));
      i++;
      continue;
    }

    const bullet = t.match(/^[-*]\s+(.*)$/);
    if (bullet) {
      blocks.push({
        object: "block",
        type: "bulleted_list_item",
        bulleted_list_item: { rich_text: richText(clean(bullet[1])) },
      });
      i++;
      continue;
    }

    const num = t.match(/^\d+\.\s+(.*)$/);
    if (num) {
      blocks.push({
        object: "block",
        type: "numbered_list_item",
        numbered_list_item: { rich_text: richText(clean(num[1])) },
      });
      i++;
      continue;
    }

    if (t.startsWith(">")) {
      blocks.push({
        object: "block",
        type: "quote",
        quote: { rich_text: richText(clean(t.replace(/^>\s?/, ""))) },
      });
      i++;
      continue;
    }

    blocks.push(para(clean(t)));
    i++;
  }
  return blocks;
}

async function notion(token: string, path: string, method: string, body: any) {
  const res = await fetch(`https://api.notion.com/v1${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${token}`,
      "Notion-Version": NOTION_VERSION,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });
  const json = await res.json();
  if (!res.ok) throw new Error(`Notion ${method} ${path} -> ${res.status}: ${JSON.stringify(json).slice(0, 400)}`);
  return json as any;
}

const file = arg("file");
const parent = arg("parent");
const title = arg("title");
if (!file || !parent || !title) {
  console.error("usage: --file <md> --parent <page_id> --title <title>");
  process.exit(1);
}

const token = await resolveToken();
const md = await Bun.file(file).text();
const blocks = mdToBlocks(md);

const page = await notion(token, "/pages", "POST", {
  parent: { page_id: parent },
  properties: { title: { title: [{ type: "text", text: { content: title } }] } },
  children: blocks.slice(0, MAX_BLOCKS_PER_CALL),
});

let appended = Math.min(blocks.length, MAX_BLOCKS_PER_CALL);
while (appended < blocks.length) {
  const chunk = blocks.slice(appended, appended + MAX_BLOCKS_PER_CALL);
  await notion(token, `/blocks/${page.id}/children`, "PATCH", { children: chunk });
  appended += chunk.length;
}

console.log(JSON.stringify({ ok: true, page_id: page.id, url: page.url, blocks: blocks.length }, null, 2));
